Use modern packaging standards for setuptools

eli-schwartz opened the pull request:
In setuptools 30.3.0 (December 2016), support was added for specifying setup options in setup.cfg as opposed to arguments to the setup function. Options are merged, and setup.py takes priority when disagreements occur.

setup.cfg arguments are superior for a couple reasons. They are:
- declarative -- no python code is executed to evaluate them
- easier to read
- capable of doing things that you'd need to open-code in python

Of particular note, it is possible to load files or statically parse the AST for a simple assignment, via values such as `attr:` or `file:`. This is a cfg-exclusive feature.

eli-schwartz committed:
Use modern packaging standards for setuptools

In setuptools 30.3.0 (December 2016), support was added for specifying
setup options in setup.cfg as opposed to arguments to the setup
function. Options are merged, and setup.py takes priority when
disagreements occur.

setup.cfg arguments are superior for a couple reasons. They are:
- declarative -- no python code is executed to evaluate them
- easier to read
- capable of doing things that you'd need to open-code in python

Of particular note, it is possible to load files or statically parse the
AST for a simple assignment, via values such as `attr:` or `file:`. This
is a cfg-exclusive feature.

eli-schwartz commented:
@jakevdp

from https://github.com/google/jax/issues/15256#issuecomment-1487730643 

> but will be somewhat painful. Our current setup process makes use of a lot of the imperative computation flexibility of `setup.py` files

Here's an example that shows you can have the best of both worlds. :)

EpicWink commented:
[pyproject.toml](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/) is the community standard for specifying project metadata declaritively: I would recommend using that over [setup.cfg](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html), which will [eventually be deprecated](https://github.com/pypa/setuptools/issues/3214) and removed.

In this case, you would have to add `"optional-dependencies"` to [`dynamic`](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#dynamic). All other features are supported in pyproject.toml, either in the `project` table or [`tool.setuptools`](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration). `url` will become [`project.urls.homepage`](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#urls).

eli-schwartz commented:
> or [`tool.setuptools`](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration).

... which is still in beta stage, unlike something that has worked robustly and stably for half a decade plus.

> over [setup.cfg](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html), which will [eventually be deprecated](https://github.com/pypa/setuptools/issues/3214) and removed.

This will almost certainly never ever ever happen, because it would cause large chunks of PyPI to become unbuildable without any hope of vanished maintainers ever fixing things. Additionally, I seem to recall the proposed "deprecation" method was actually "internally run a function that converts cfg to toml, and then run another function to load the toml", which is a bit of indirection that really boils down to "setup.cfg is still supported".

EpicWink commented:
`find_packages` is now unused in setup.py: I suggest removing the import

---

> > or [tool.setuptools](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration).
>
> ... which is still in beta stage

The docs aren't particularly clear, but it seems specifically automatic discovery, package data, and dynamic dependencies (from file) are in beta. I suppose its fine to wait for it to be marked as stable.

---

> > over [setup.cfg](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html), which will https://github.com/pypa/setuptools/issues/3214 and removed.
>
> This will almost certainly never ever ever happen, because it would cause large chunks of PyPI to become unbuildable without any hope of vanished maintainers ever fixing things.

Thus the deprecation. As a related example, `distutils` is in the final stages of being removed, despite it being used in ~100% of distributed projects 10 or 15 years ago.

eli-schwartz committed:
Migrate off of deprecated distutils APIs

The find_executable() function is pretty nonstandard and obscure, and
distutils itself is deprecated in CPython and undergoing rapid evolution
in setuptools with an eye toward getting rid of APIs deemed not useful
or well used.

Since python 3.3, shutil.which is the popular way to do this now, anyway.

eli-schwartz commented:
> `find_packages` is now unused in setup.py: I suggest removing the import

Good point. While I'm at it, distutils.spawn.find_executable() is a truly old way of performing the role of shutil.which... speaking of deprecated things.
